File: apps/api/src/votes/entities/vote.entity.ts
import { Entity, Column, PrimaryGeneratedColumn, ManyToOne, JoinColumn, Index } from 'typeorm';
import { Account } from 'src/account/entities/account.entity';
import { EntityType } from 'src/common/enums/entity-type.enum';
import { VoteType } from 'src/common/enums/vote-type.enum';

@Entity({ name: 'votes' })
@Index(['userId', 'entityType', 'entityId'], { unique: true })
export class Vote {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'user_id', nullable: false })
  userId: number;

  @ManyToOne(() => Account)
  @JoinColumn({ name: 'user_id' })
  user: Account;

  @Column({ name: 'entity_id', type: 'bigint', nullable: false })
  entityId: number;

  @Column({ name: 'entity_type', type: 'enum', enum: EntityType, nullable: false })
  entityType: EntityType;

  @Column({ name: 'vote_type', type: 'enum', enum: VoteType, nullable: false })
  voteType: VoteType;

  @Column({ name: 'created_at', type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
  createdAt: Date;
}

File: apps/api_go/internal/domain/vote_entity.go
package domain

import "gorm.io/gorm"

type Vote struct {
	gorm.Model
	UserID     uint       `gorm:"column:user_id;not null"`
	User       *Account   `gorm:"foreignKey:UserID"`
	EntityID   int64      `gorm:"column:entity_id;type:bigint;not null"`
	EntityType EntityType `gorm:"column:entity_type;type:varchar(20);not null"`
	VoteType   VoteType   `gorm:"column:vote_type;type:varchar(10);not null"`
}

func (Vote) TableName() string {
	return "votes"
}
